### PR TITLE
Reorganize AST nodes related to size and shape of arrays

### LINKED ISSUE #1329: Rename `NumpyArraySize`
As discussed in https://github.com/pyccel/pyccel/pull/1326#discussion_r1112041778, the name `NumpyArraySize` is misleading as it does not return the same thing as NumPy's `size` function. We should therefore rename the classes `NumpyArraySize` and `PyccelArraySize` as follows:

`NumpyArraySize`->`NumpyArrayShapeElement`
`PyccelArraySize`->`NumpyArraySize`

### PR DESCRIPTION
**Pull request by yguclu**
Fix #1329: rename and reorganize classes (AST nodes) which compute the size and shape of arrays.

List of changes
---------------

- Rename `Shape` as `NumpyShape` in ast.numpyext module
- Rename `NumpyArraySize` as `NumpySize` in ast.numpyext module
- Rename `PyccelArraySize` as `PyccelArrayShapeElement` in ast.internals
- Create new class `PyccelArraySize` without `index` argument
- Make `NumpySize` a factory class (which cannot be instantiated)
- Add unit tests for `axis` argument to `NumpySize`

### CODE REVIEW

**Comment by yguclu** [comment]
@EmilyBourne Regarding commit 7265bd3 the import of `IndexedElement` was missing in the file. This means that the [final return statement](https://github.com/pyccel/pyccel/blob/7265bd3c9cf0fefcdac8613fa78d58d694b9fc2a/pyccel/ast/numpyext.py#LL1708C3-L1708C51) of `NumpySize.__new__` is never run in our tests:
```python
    def __new__(cls, a, axis = None):

        if axis is None:
            return PyccelArraySize(a)

        if isinstance(axis, LiteralInteger) and a.shape is not None:
            return a.shape[axis.python_value]

        return IndexedElement(NumpyShape(a), axis)
```

**Comment by EmilyBourne** [comment]
> @EmilyBourne Regarding commit [7265bd3](https://github.com/pyccel/pyccel/commit/7265bd3c9cf0fefcdac8613fa78d58d694b9fc2a) the import of `IndexedElement` was missing in the file. This means that the [final return statement](https://github.com/pyccel/pyccel/blob/7265bd3c9cf0fefcdac8613fa78d58d694b9fc2a/pyccel/ast/numpyext.py#LL1708C3-L1708C51) of `NumpySize.__new__` is never run in our tests:
> 
> ```python
>     def __new__(cls, a, axis = None):
> 
>         if axis is None:
>             return PyccelArraySize(a)
> 
>         if isinstance(axis, LiteralInteger) and a.shape is not None:
>             return a.shape[axis.python_value]
> 
>         return IndexedElement(NumpyShape(a), axis)
> ```

Did you collect the tests from #1387 ? Otherwise there are no tests for the `axis` argument

**Comment by yguclu** [comment]
@EmilyBourne I had completely forgotten about this PR :sweat_smile: Should we try to merge it now?

**Comment by yguclu** [comment]
It appears that we must add lots of docstrings before merging this PR :rofl:

File: pyccel/ast/internals.py [hunk 1]
```diff
@@ -55,88 +56,132 @@
...
+        """
+        return self._args[0]
+
+    def __str__(self):
+        return f'Size({self.arg})'
+
+    def __eq__(self, other):
+        if isinstance(other, PyccelArraySize):
+            return self.arg == other.arg
+        else:
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/internals.py:113]: @yguclu Do you think we should ignore this? It is for comparing sizes. I think it was used to check if arrays were compatible but the handling of this regressed a little when we added functions returning arrays. I think it could be useful in the future and is simple enough to not be critical

File: pyccel/ast/numpyext.py [hunk 2]
```diff
@@ -1679,52 +1676,33 @@
...
+        if axis is None:
+            return PyccelArraySize(a)
 
-    @property
-    def arg(self):
-        """ Object whose size is investigated
-        """
-        return self._arg
+        if isinstance(axis, LiteralInteger) and a.shape is not None:
+            return a.shape[axis.python_value]
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:1859]: I am surprised that this is never used. Have you looked into it at all?

File: pyccel/ast/numpyext.py [hunk 3]
```diff
@@ -1679,52 +1820,44 @@ def keep_dims(self):
...
 
-class NumpyArraySize(PyccelInternalFunction):
+
+class NumpySize(PyccelInternalFunction):
     """
-    Class representing a call to the numpy size function which
-    returns the shape of an object in a given dimension
+    Represent a call to numpy.size in the user code.
+
+    This wrapper class represents a call to the Numpy size function, which
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:1828]:
```suggestion
    This wrapper class represents a call to the NumPy `size` function, which
```

File: pyccel/ast/numpyext.py [hunk 4]
```diff
@@ -1679,52 +1820,44 @@ def keep_dims(self):
...
+    returns the total number of elements in a multidimensional array, or the
+    number of elements along a given dimension.
+
+    Objects of this class are never present in the Pyccel AST, because the
+    class constructor returns objects of type `PyccelArraySize`, `Literal`, or
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:1833]:
```suggestion
    class constructor returns objects of type `PyccelArraySize`, `LiteralInteger`, or
```

File: pyccel/ast/numpyext.py [hunk 5]
```diff
@@ -1679,52 +1820,44 @@ def keep_dims(self):
...
-    axis  : int
-            The dimension along which the size is
-            requested
+    ----------
+    a : PyccelAstNode
+        An array of unknown size.
+
+    axis : int, optional
+        The dimension along which the size is requested.
+
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:1843]: Looking at the code it seems that `axis` can also be a `Variable` or a `LiteralInteger` (or maybe an expression?). Maybe the more general `PyccelAstNode` would be more accurate. `int` can be kept if it is used as a shortcut in code written by developers, but I am not sure that this is useful for `NumpySize`

**Comment by yguclu** [comment]
@EmilyBourne Can we just ignore the two coverage errors?

File: pyccel/ast/variable.py [hunk 6]
```diff
@@ -219,49 +219,72 @@ def __init__(
...
     def process_shape(self, shape):
-        """ Simplify the provided shape and ensure it
-        has the expected format
+        """
+        Simplify the provided shape and ensure it has the expected format.
+
+        The provided shape is the shape used to create the object. In most
+        cases where the shape is required we do not require this expression
+        (which can be quite long). This function therefore replaces those
+        expressions with calls to PyccelArrayShapeElement.
```
Comment (EmilyBourne) [inline pyccel/ast/variable.py:228]: I think it is worth mentioning that the original expression is not only removed in case it is long but also because it can become invalid. Eg:
```python
n = 2
x = np.ones(n) # _shape = Variable('n') ok here
n = 3
print(x.shape[0]) # _shape = Variable('n') wrong here
```

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 7]
```diff
@@ -1241,16 +1241,29 @@ def test_size_1d(f):
...
 
+    @types('int[:,:]','int')
+    def test_size_axis(f, axis):
+        from numpy import size
+        return size(f,axis)
+
+
     from numpy import empty
-    f1 = epyccel(test_size_1d, language = language)
+    f1 = epyccel(test_size_1d ,language = language)
```
Comment (EmilyBourne) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:1251]: Was this change intentional?

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
This looks good but I have a few comments

**Comment by EmilyBourne** [comment]
> @EmilyBourne Can we just ignore the two coverage errors?

I'm not sure if you are asking about the mechanics of the coverage bot (in which case see the message left by the bot), or for my opinion on the 2 untested code sections (in which case see the comments I left 30mins ago on the relevant conversations)

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 7 above)
Comment (yguclu) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py:1251]: Did I really make this change? 😅

File: pyccel/ast/internals.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/ast/internals.py:113]: I am OK with ignoring this warning 😉

File: pyccel/ast/numpyext.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/ast/numpyext.py:1859]: I have not

File: pyccel/ast/internals.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/internals.py:113]: /bot accept for reasons described above

File: CHANGELOG.md [hunk 8]
```diff
@@ -59,6 +59,9 @@ All notable changes to this project will be documented in this file.
 
 -   #1355 : Remove unused `BasicParser` arguments.
 -   \[INTERNALS\] Re-write bind-c wrapping stage (#1388)
+-   \[INTERNALS\] In `ast.numpyext` rename `Shape` as `NumpyShape`, `NumpyArraySize` as `NumpySize`
+-   \[INTERNALS\] In `ast.internals` rename `NumpyArraySize` as `NumpySize`, create new `NumpyArraySize` w/out `index` argument
+-   \[INTERNALS\] Make `NumpySize` a factory class (which cannot be instantiated)
```
[the remark this answers is not in the document]
Comment (yguclu) [reply CHANGELOG.md]: Fixed in a3fe7d7.

File: pyccel/ast/internals.py [hunk 9]
```diff
@@ -29,100 +30,182 @@
...
         return self._args
 
     @property
     def is_elemental(self):
-        """ Indicates whether the function should be
-        called elementwise for an array argument
+        """
+        Whether the function acts elementwise on an array argument.
+
+        Bool indicating whether the (scalar) function should be called
```
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/ast/internals.py]: Fixed in aa20b03.

File: pyccel/ast/internals.py [hunk 10]
```diff
@@ -29,100 +30,182 @@
...
     _rank  = 0
     _shape = None
     _order = None
 
     def __init__(self, arg, index):
-        if not isinstance(arg, (list,
-                                tuple,
-                                PyccelAstNode)):
-            raise TypeError('Unknown type of  %s.' % type(arg))
+        if not isinstance(arg, (list, tuple, PyccelAstNode)):
```
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/ast/internals.py:142]: I do not know. This was in the original code, but it is inconsistent with the docstring. I can remove those options and see if the tests still pass.

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 7 above)
Comment (yguclu) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: Fixed in ca9e379

File: pyccel/ast/numpyext.py (hunk 5 above)
Comment (yguclu) [reply pyccel/ast/numpyext.py]: Fixed in c82512e

File: pyccel/ast/variable.py (hunk 6 above)
Comment (yguclu) [reply pyccel/ast/variable.py]: Fixed in fcfd775

File: pyccel/ast/internals.py [hunk 11]
```diff
@@ -29,100 +30,182 @@
...
+    Class representing a call to a function which would return
+    the shape of a multi-dimensional array in a given dimension.
 
     Parameters
-    ==========
-    arg   : PyccelAstNode
-            A PyccelAstNode of unknown shape
+    ----------
+    arg : PyccelAstNode
+        An array of unknown shape.
```
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/ast/internals.py:127]: See answer above

File: pyccel/ast/internals.py [hunk 12]
```diff
@@ -29,100 +30,182 @@
...
 
+
 class Slice(Basic):
+    """
+    Represents a slice in the code.
 
-    """Represents a slice in the code.
+    Objects of this class represent the slicing of a Numpy array along one of
+    its dimensions, and they are mostly created during the syntactic stage
+    from objects of the corresponding class `python.ast.Slice`.
```
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/ast/internals.py]: You are correct. I improved the docstring in 97abb20

File: pyccel/ast/internals.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (yguclu) [reply pyccel/ast/internals.py]: See a97950d

File: pyccel/ast/numpyext.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:1850]: I have had a quick look. This is not used because there is only one test for `size` with the axis argument and it does not use a literal argument. I think a second test should be added to cover this case

**Review by EmilyBourne — APPROVED** [review]
